mscore/resourceManager.cpp

```diff
@@ -373,6 +860,228 @@ void ResourceManager::downloadExtension()
+            refreshPluginButton(button->property("row").toInt());
+            writePluginPackages();
+            displayPlugins();
+            }
+      else {
+            button->setText("Bad archive. Try again");
+            button->setEnabled(true);
+            return;
+            }
+      }
+
+void ResourceManager::uninstallPluginPackage() 
+      {
+      QPushButton* button = static_cast<QPushButton*>(sender());
+      const QString& url = button->property("page_url").toString();
+      if (!pluginDescriptionMap.contains(url))
+            return;
+      PluginPackageDescription& desc = pluginDescriptionMap[url];
+      // remove qml files first(which may be outside the folder of this plugin)
+      for (QString& path : desc.qml_paths)
+            QFile::remove(path);
+      // then remove the folder
+      QDir d(desc.dir);
+      if (!d.removeRecursively()) {
+            button->setText("Failed, try again");
+            return;
+            }
+      pluginDescriptionMap.remove(url);
+      refreshPluginButton(button->property("row").toInt());
+      writePluginPackages();
+      displayPlugins();
+      }
+
+
+void ResourceManager::refreshPluginButton(int row, bool updated/* = true*/)
+      {
+      // TODO: get button, then plugin url, then update status
+      // install button
+      QPushButton* install = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(row, 3)));
+      QPushButton* uninstall = static_cast<QPushButton*>(pluginsTable->indexWidget(pluginsTable->model()->index(row, 4)));
+      const QString& page_url = install->property("page_url").toString();
+      bool installed = pluginDescriptionMap.contains(page_url);
+      uninstall->setEnabled(installed);
+      if (installed) {
+            if (updated) {
+                  install->setText("Updated");
+                  install->setEnabled(false);
+                  }
+            else {
+                  install->setText("Update");
+                  install->setEnabled(true);
+                  connect(install, SIGNAL(clicked()), this, SLOT(updatePlugin()));
+                  }
+            }
+      else {
+            install->setText("Install");
+            install->setEnabled(true);
+            connect(install, SIGNAL(clicked()), this, SLOT(downloadInstallPlugin()));
+            }
+
+      }
+
+//---------------------------------------------------------
+//   writePluginPackages
+//---------------------------------------------------------
+void ResourceManager::writePluginPackages()
+      {
+      QDir dir;
+      dir.mkpath(dataPath);
+      QFile f(dataPath + "/pluginpackages.xml");
+      if (!f.open(QIODevice::WriteOnly)) {
+            qDebug("cannot create plugin package file <%s>", qPrintable(f.fileName()));
+            return;
+            }
+      XmlWriter xml(0, &f);
+      xml.header();
+      xml.stag("museScore version=\"" MSC_VERSION "\"");
+      foreach(const QString &pkg, pluginDescriptionMap.keys()) {
```

Comment:
likewise

Reply:
Resolved in 77701599820befd5e5381af2075d63e6b5b3dd85.